### PR TITLE
Batching rule for Tensor.new_empty_strided

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47227 Batched gradient support for view+inplace operations
* #47365 Add batching rule for torch.clone(tensor, torch.contiguous_format)
* #47364 as_strided batching rule
* **#47226 Batching rule for Tensor.new_empty_strided**
* #47225 Implement Tensor.new_empty_strided(sizes, strides, *, dtype, device, requires_grad)

The batching rule is a little weird because it's not immediately obvious
what the strides of the result should be. If
tensor.new_empty_strided(size, stride) is called inside vmap and
`tensor` is being vmapped over, the result is a physical tensor with:
- size `[batch_shape] + size`
- strides `[S0, S1, ..., Sn] + stride` such that the
S0...Sn are part of a contiguous subspace and Sn is equal to the size of
the storage of `torch.empty_strided(size, stride)`.

I refactored some of the logic that computes the storage size for
`torch.empty_strided(size, stride)` into a helper function
`native::storage_size_for` and use it in the batching rule.

Test Plan:
- New tests in test/test_vmap.py

Differential Revision: [D24741690](https://our.internmc.facebook.com/intern/diff/D24741690)

### CODE REVIEW

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 1]
```diff
@@ -694,6 +695,53 @@ Tensor new_empty_batching_rule(
...
+  // shape of the result should be ([B0, B1, B2] + size), but what about
+  // the physical strides?
+  //
+  // We're actually free to pick whatever stride we want:
+  // e.g., for size=[5, 3], stride=[0, 1], we could decide to
+  // use
+  // - physical size: [B0, B1, B2, 5, 3]
+  // - physical stride: [100*B1*B2, 100*B2, 100, 0, 1]
+  //
+  // Let's select the strides that result in the smallest memory usage.
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:716]: I am not sure about this argument. Even we we transposed everything to be "column contiguous" the actual memory usage would be the same no?
Is the choice here done to make the batch tensor contiguous if the original result of `empty_strided(size, stride)` was contiguous?

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 2]
```diff
@@ -694,6 +695,53 @@ Tensor new_empty_batching_rule(
...
 
+Tensor new_empty_strided_batching_rule(
+    const Tensor& self,
+    IntArrayRef size,
+    IntArrayRef stride,
+    const TensorOptions& options) {
+  auto physical_view = MultiBatchVmapTransform::logicalToPhysical(self);
+  auto physical_size = physical_view.getPhysicalShape(size);
+
+  // Let [B0, B1, B2] be the shape of the batch dims. We know what the physical
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:706]: What happens if all the batch dimensions are not at the front?

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 3]
```diff
@@ -694,6 +695,53 @@ Tensor new_empty_batching_rule(
...
+  auto batch_shape = IntArrayRef(
+      physical_view.tensor().sizes().begin(), physical_view.numBatchDims());
+
+  // physical_strides = [B1 * B2 * S, B2 * S, S]
+  auto physical_strides = at::detail::defaultStrides(batch_shape);
+  TORCH_CHECK(size.size() == stride.size(),
+        "new_empty_strided(sizes, strides): dimensionality of sizes (",
+        size.size(), ") must match dimensionality of strides (",
+        stride.size(), ")");
+  auto example_storage_size = native::storage_size_for(size, stride);
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:731]: nit: why is this one called "example"?

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 4]
```diff
@@ -694,6 +695,53 @@ Tensor new_empty_batching_rule(
...
+  for (int64_t idx = 0; idx < physical_strides.size(); ++idx) {
+    physical_strides[idx] *= example_storage_size;
+  }
+
+  // physical_strides = [B1 * B2 * S, B2 * S, S] + strides
+  physical_strides.insert(physical_strides.end(), stride.begin(), stride.end());
+
+  auto result = physical_view.tensor().new_empty_strided(
```
Comment (albanD) [inline aten/src/ATen/BatchingRegistrations.cpp:739]: nit: Do you have to call back into `new_empty_strided` here? Or could it be done with `at::empty_strided` and this is just for consistency?

File: aten/src/ATen/native/Resize.h [hunk 5]
```diff
@@ -51,15 +52,7 @@ inline TensorImpl* resize_impl_cpu_(
   if (stride) {
     self->set_sizes_and_strides(size, *stride);
     // NB: storage size can be different from numel.
-    for (size_t dim = 0; dim < size.size(); ++dim) {
-      // FIXME: Don't rely on storage_size being negative because this
```
Comment (albanD) [inline aten/src/ATen/native/Resize.h:55]: I see that this comment was dropped. Do you know what it was referring to?

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:731]: MultiBatchVmapTransform::logicalToPhysical(self) moved all of the batch dims to the front

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:741]: If we picked `[100*B1*B2, 100*B2, 100, 0, 1]` instead of `[B1 * B2 *S, B2 * S, S, 5, 3]`, then the resulting storage size would be very large (100*B1*B2 * (B0-1) would be the memory location of result[B0-1], rather than B1 * B2 *S * (B0-1))

If the original result of `empty_strided(size, stride)` is contiguous, then the batched tensor is contiguous, yes -- that is also one of the motivations (I'll add that to the comment)

What do you mean by "column contiguous"?

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:731]: But I should clarify in the comment that we are going to create the batch dimensions at the front, even if in the original tensors being vmapped the batch dim wasn't at the front.

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:764]: This can be done with at::empty_strided. I chose to use `new_empty_strided` because it was the first thing that came to mind. I like at::empty_strided because it's less verbose, let's go with that.

File: aten/src/ATen/native/Resize.h (hunk 5 above)
Comment (zou3519) [reply aten/src/ATen/native/Resize.h:55]: I wrote this comment two years ago but unfortunately I don't remember why :/

Looking at [the original PR](https://github.com/pytorch/pytorch/pull/13185/files#diff-67630b8526b429e0fc2c7851f15eb290808c90d898e8ab98fb28821d513c3a46R38), previously, storage_size could be negative if size[dim] is equal to 0. I think that is what the comment is referring to.

However, in the current iteration of the code, storage_size can never be negative, so I think the comment doesn't apply anymore.

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:764]: Actually, I take that back: there is indeed a semantic difference. If the user didn't provide some of the options, then we use the options of the original tensor. For example:

```
x = torch.rand(3, device='cuda')
y = x.new_empty_strided([4, 4], [4, 1])
```
this should create a new CUDA tensor. However, none of the options being passed to `new_empty_strided_batching_rule` have actually been set. So if we used `at::empty(physical_size, physical_strides, options)`, this would end up creating a CPU tensor (the default for device when it is not set is CPU)

File: aten/src/ATen/native/Resize.h (hunk 5 above)
Comment (albanD) [reply aten/src/ATen/native/Resize.h:55]: Right the dim=0 case is special cased so that should be fine!

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 1 above)
Comment (albanD) [reply aten/src/ATen/BatchingRegistrations.cpp:741]: > What do you mean by "column contiguous"?

Sorry that is not very clear. Stride being strictly increasing. Matching a column-major packing.

> If we picked [100*B1*B2, 100*B2, 100, 0, 1] instead of [B1 * B2 *S, B2 * S, S, 5, 3]

You mean using another fixed values V (=100 here) such that S <= V?
I guess that is true but I'm not sure why would anyone argue that this is a thing we want to do?

Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:741]: > You mean using another fixed values V (=100 here) such that S <= V?
I guess that is true but I'm not sure why would anyone argue that this is a thing we want to do?

Yes, exactly. It's a degree of freedom that I was shutting down for completeness. I'll refactor the comment to make it clear that the reason is more to make the batch tensor contiguous if the original result of empty_strided(size, stride) is contiguous (and if the original result is not contiguous, for the batch dimensions to be contiguous in memory)

**Review by albanD — APPROVED** [review]
lgtm
